### DISCUSSION TITLE
Joining Requests and Responses

### DISCUSSION THREAD

**Discussion by hillaryan** [cesanta/mongoose#2426] [Ask questions, get answers] [answered]
Good Afternoon,

My understanding of the general setup for an http client with mongoose looks like this:

- To connect to a server, I first call `mg_http_connect` with a callback to handle events.
- Once the callback registers the `MG_EV_CONNECT` event, I am ready to start sending requests via `mg_printf`.
- When the server replies to these requests, I can read the replies in the callback by watching for the `MG_EV_HTTP_MSG` event.

So far, so good, but how do I correlate requests with responses in a connection that is kept alive? 

*Addendum:* I only have a superficial understanding of the HTTP protocol and how it has evolved over time. It seems to me that mongoose only supports HTTP1.0 & HTTP1.1. In these protocols, unlike HTTP2.0, it looks like requests must be responded to in the order that they are received. 

Given the flexibility of the mongoose server setup, it seems quite possible to reply to messages out of order. Assuming this is correct and given that latency is not a big issue, it seems that the easiest way to handle this limitation is to close connections after a transaction has been completed.

**Comment by scaprile** [comment, marked as answer]
The same way you do it with any other client, you send the request, you receive the response

https://mongoose.ws/documentation/tutorials/http-server/#how-to-get-request-data

Please follow the guidelines in the [documentation](https://mongoose.ws/documentation), [examples](https://github.com/cesanta/mongoose/tree/master/examples) and [tutorials](https://mongoose.ws/tutorials/) available.
